src/Components/Analyzers/src/ComponentParameterUsageAnalyzer.cs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Components.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class ComponentParameterUsageAnalyzer : DiagnosticAnalyzer
+    {
+        public ComponentParameterUsageAnalyzer()
+        {
+            SupportedDiagnostics = ImmutableArray.Create(new[]
+            {
+                DiagnosticDescriptors.ComponentParametersShouldNotBeSetOutsideOfComponents,
+            });
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!ComponentSymbols.TryCreate(context.Compilation, out var symbols))
+                {
+                    // Types we need are not defined.
+                    return;
+                }
+
+                context.RegisterOperationBlockStartAction(startBlockContext =>
```

Comment:
Is there a good reason why this needs to be split into start/end instead of just being a simple operation action?

It looks like this goes block by block to collection the references, and then loops through them later, so it seems like it could be simplified to just do the work inside `RegisterOperationAction`?

Reply:
It could, I was just following the pattern of finding subjects and then possibly logging errors for them in the end action.